src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -150,31 +135,159 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
             if (dimensions > 1)
-                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(TAnyElement).Name}>");
+                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(TRequestedElement).Name}>");
+
+            var containsNulls = buf.ReadInt32() == 1;
+            if (!ElementTypeInfo<TRequestedElement>.IsReferenceOrNullable && containsNulls)
+                throw new InvalidOperationException(
+                    $"Can't read a non-nullable List<{typeof(TRequestedElement).Name}> if the database array field contains null values. Use List<{typeof(TRequestedElement).Name}?> instead.");
 
-            buf.ReadInt32();  // Has nulls. Not populated by PG?
             buf.ReadUInt32(); // Element OID. Ignored.
 
             await buf.Ensure(8, async);
             var length = buf.ReadInt32();
             buf.ReadInt32(); // We don't care about the lower bounds
 
-            var list = new List<TAnyElement>(length);
+            var list = new List<TRequestedElement>(length);
             for (var i = 0; i < length; i++)
-                list.Add(await _elementHandler.ReadWithLength<TAnyElement>(buf, async));
+                list.Add(await ElementHandler.ReadWithLength<TRequestedElement>(buf, async));
             return list;
         }
 
+        #endregion Read
+
+        #region Static generic caching helpers
+
+        internal static class ElementTypeInfo<TElement>
+        {
+            public static readonly bool IsReferenceOrNullable =
+                !typeof(TElement).IsValueType ||
+                typeof(TElement).IsGenericType && typeof(TElement).GetGenericTypeDefinition() == typeof(Nullable<>);
+
+            public static readonly List<TElement> EmptyList = new List<TElement>(0);
+            // ReSharper disable StaticMemberInGenericType
+            public static readonly Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<Array>> ReadNullableArrayFunc = default!;
+            // ReSharper restore StaticMemberInGenericType
+
+            static ElementTypeInfo()
+            {
+                if (IsReferenceOrNullable)
+                    return;
+
+                var arrayHandlerParam = Expression.Parameter(typeof(ArrayHandler), "arrayHandler");
+                var bufferParam = Expression.Parameter(typeof(NpgsqlReadBuffer), "buf");
+                var asyncParam = Expression.Parameter(typeof(bool), "async");
+
+                ReadNullableArrayFunc = Expression.Lambda<Func<ArrayHandler, NpgsqlReadBuffer, bool, ValueTask<Array>>>(
+                        Expression.Call(
+                            arrayHandlerParam,
+                            ReadArrayMethod.MakeGenericMethod(
+                                typeof(Nullable<>).MakeGenericType(typeof(TElement))),
+                            bufferParam, asyncParam),
+                        arrayHandlerParam, bufferParam, asyncParam)
+                    .Compile();
```

Comment:
That's not possible, because, as I wrote above, `ElementTypeInfo<TElement>` is nested inside `ArrayHandler` not inside `ArrayHandler<TElement>` so in this case `TElement` might be the element type of any array that the consumer requested.

Reply:
> so in this case `TElement` might be the element type of any array that the consumer requested.

It's even worse, as in this case `TElement` isn't even available as we're inside `ArrayHandler` not inside `ArrayHandler<TElement>` which is intentional as we want to support both `TElement` and `TElement?`.

Reply:
Yep, my mistake. There is a little difference between methods I lost, but hen there is a question. What if an array must have no nulls in it because its elements are `NOT NULL` domains? Right now we a going to force users to use array with nullable elements, this is a breakage. At the same time it will be published in v5.0, so it's okay probably.

@austindrenski @roji, your thoughts?

Reply:
What I asked first is possible, but it's too tricky and introduce more code than your solution. Let's keep it as is in that case (:

Reply:
Good question.
I've actually never thought about domains with a `NOT NULL` constraint as a way to avoid `NULL` values inside an array.
While trying to look up if this is even supported I found out that the use of `NOT NULL` constraints on domains is [mildly discouraged](https://www.postgresql.org/docs/current/sql-createdomain.html#id-1.9.3.62.7) because it has some tricky corner cases but I've confirmed that it is possible to create and use arrays of non-nullable domains, which renders your question perfectly valid.
On the other hand we're probably talking about a corner case of a corner case feature and we only force them if they use `NpgsqlDataReader.GetValue()`. If they use `NpgsqlDataReader.GetFieldValue<T>()` they can always request an array of non-nullable value types.

Reply:
I really like the idea of `NOT NULL` domains - good idea @YohDeadfall. Even if `NOT NULL` has its caveats, it's very clear information on the type expressing the intention for it never to contain nulls.

I agree with @Brar that it's a corner case, but I still do think it would be nice to do, at least for completeness. It also shouldn't be very hard: [pg_type contains `typnotnull`](https://www.postgresql.org/docs/current/catalog-pg-type.html) which could be loaded and stored on PostgresDomainType, and we could check that from ArrayHandler.GetValue.

But we shouldn't block this PR on this (unless you really feel like doing this right away @Brar). Unless someone objects I can open a separate issue to traack this.

Reply:
I'm okay with doing this in the next iteration, @Brar.

Reply:
I'm all for the separate issue.
I also feel like it's not too complicated, so we can possibly get this done for 5.0 but I really want to avoid feature creep for this PR as it has already taken long enough to bring it to a state that we all can possibly agree on.
I'll have a look at it after we've merged this, if you want me to.

Reply:
Opened #2720 to track, @Brar you're welcome to take this if you're interested.